### PR TITLE
Add support for sending files while editing messages

### PR DESCRIPTION
**Pull request by IDoEverything**
# Summary
Adds the ability to send files while editing messages.

# Details
Changed `EditMessageAsync` and `EditWebhookMessageAsync` to use multipart to be able to send files.

### CODE REVIEW

File: DSharpPlus/Net/Rest/DiscordApiClient.cs [hunk 1]
```diff
@@ -1063,14 +1063,24 @@ internal async Task<DiscordMessage> EditMessageAsync(ulong channel_id, ulong mes
...
+            {
+                ["payload_json"] = DiscordJson.SerializeObject(pld)
+            };
+
             var route = $"{Endpoints.CHANNELS}/:channel_id{Endpoints.MESSAGES}/:message_id";
             var bucket = this.Rest.GetBucket(RestRequestMethod.PATCH, route, new { channel_id, message_id }, out var path);
 
             var url = Utilities.GetApiUriFor(path);
-            var res = await this.DoRequestAsync(this.Discord, bucket, url, RestRequestMethod.PATCH, route, payload: DiscordJson.SerializeObject(pld)).ConfigureAwait(false);
+            var res = await this.DoMultipartAsync(this.Discord, bucket, url, RestRequestMethod.PATCH, route, values: values, files: files ?? new List<DiscordMessageFile>()).ConfigureAwait(false);
```
Comment (VelvetThePanda) [inline DSharpPlus/Net/Rest/DiscordApiClient.cs:1075]: Don't do this

File: DSharpPlus/Net/Rest/DiscordApiClient.cs [hunk 2]
```diff
@@ -1044,7 +1044,7 @@ internal async Task<DiscordMessage> GetChannelMessageAsync(ulong channel_id, ulo
             return ret;
         }
 
-        internal async Task<DiscordMessage> EditMessageAsync(ulong channel_id, ulong message_id, Optional<string> content, Optional<IEnumerable<DiscordEmbed>> embeds, IEnumerable<IMention> mentions, IReadOnlyList<DiscordActionRowComponent> components)
+        internal async Task<DiscordMessage> EditMessageAsync(ulong channel_id, ulong message_id, Optional<string> content, Optional<IEnumerable<DiscordEmbed>> embeds, IEnumerable<IMention> mentions, IReadOnlyList<DiscordActionRowComponent> components, IReadOnlyCollection<DiscordMessageFile> files)
```
Comment (VelvetThePanda) [inline DSharpPlus/Net/Rest/DiscordApiClient.cs:1047]: Why not IEnumerable

File: DSharpPlus/Entities/Channel/Message/DiscordMessage.cs [hunk 3]
```diff
@@ -471,7 +471,7 @@ public Task<DiscordMessage> ModifyAsync(Optional<string> content, Optional<Disco
         /// <exception cref="Exceptions.BadRequestException">Thrown when an invalid parameter was provided.</exception>
         /// <exception cref="Exceptions.ServerErrorException">Thrown when Discord is unable to process the request.</exception>
         public Task<DiscordMessage> ModifyAsync(Optional<string> content, Optional<IEnumerable<DiscordEmbed>> embeds = default)
-            => this.Discord.ApiClient.EditMessageAsync(this.ChannelId, this.Id, content, embeds, default, default);
+            => this.Discord.ApiClient.EditMessageAsync(this.ChannelId, this.Id, content, embeds, default, default, default);
```
Comment (VelvetThePanda) [inline DSharpPlus/Entities/Channel/Message/DiscordMessage.cs:474]: `Array.Empty<T>`

File: DSharpPlus/Entities/Channel/Message/DiscordMessage.cs [hunk 4]
```diff
@@ -458,7 +458,7 @@ public Task<DiscordMessage> ModifyAsync(Optional<DiscordEmbed> embed = default)
         /// <exception cref="Exceptions.BadRequestException">Thrown when an invalid parameter was provided.</exception>
         /// <exception cref="Exceptions.ServerErrorException">Thrown when Discord is unable to process the request.</exception>
         public Task<DiscordMessage> ModifyAsync(Optional<string> content, Optional<DiscordEmbed> embed = default)
-            => this.Discord.ApiClient.EditMessageAsync(this.ChannelId, this.Id, content, embed.HasValue ? new[] {embed.Value} : Array.Empty<DiscordEmbed>(), default, default);
+            => this.Discord.ApiClient.EditMessageAsync(this.ChannelId, this.Id, content, embed.HasValue ? new[] {embed.Value} : Array.Empty<DiscordEmbed>(), default, default, default);
```
Comment (VelvetThePanda) [inline DSharpPlus/Entities/Channel/Message/DiscordMessage.cs:461]: `Array.Empty<T>`

File: DSharpPlus/Entities/Channel/Message/DiscordMessage.cs [hunk 5]
```diff
@@ -445,7 +445,7 @@ public Task<DiscordMessage> ModifyAsync(Optional<string> content)
         /// <exception cref="Exceptions.BadRequestException">Thrown when an invalid parameter was provided.</exception>
         /// <exception cref="Exceptions.ServerErrorException">Thrown when Discord is unable to process the request.</exception>
         public Task<DiscordMessage> ModifyAsync(Optional<DiscordEmbed> embed = default)
-            => this.Discord.ApiClient.EditMessageAsync(this.ChannelId, this.Id, default, embed.HasValue ? new[] {embed.Value} : Array.Empty<DiscordEmbed>(), default, default);
+            => this.Discord.ApiClient.EditMessageAsync(this.ChannelId, this.Id, default, embed.HasValue ? new[] {embed.Value} : Array.Empty<DiscordEmbed>(), default, default, default);
```
Comment (VelvetThePanda) [inline DSharpPlus/Entities/Channel/Message/DiscordMessage.cs:448]: `Array.Empty<T>`

File: DSharpPlus/Entities/Channel/Message/DiscordMessage.cs [hunk 6]
```diff
@@ -433,7 +433,7 @@ internal void PopulateMentions()
         /// <exception cref="Exceptions.BadRequestException">Thrown when an invalid parameter was provided.</exception>
         /// <exception cref="Exceptions.ServerErrorException">Thrown when Discord is unable to process the request.</exception>
         public Task<DiscordMessage> ModifyAsync(Optional<string> content)
-            => this.Discord.ApiClient.EditMessageAsync(this.ChannelId, this.Id, content, default, default, default);
+            => this.Discord.ApiClient.EditMessageAsync(this.ChannelId, this.Id, content, default, default, default, default);
```
Comment (VelvetThePanda) [inline DSharpPlus/Entities/Channel/Message/DiscordMessage.cs:436]: `Array.Empty<T>`

File: DSharpPlus.Rest/DiscordRestClient.cs [hunk 7]
```diff
@@ -558,7 +558,7 @@ public Task<DiscordMessage> EditMessageAsync(ulong channel_id, ulong message_id,
         /// <param name="embed">New message embed</param>
         /// <returns></returns>
         public Task<DiscordMessage> EditMessageAsync(ulong channel_id, ulong message_id, Optional<DiscordEmbed> embed)
-            => this.ApiClient.EditMessageAsync(channel_id, message_id, default, embed.HasValue ? new[] {embed.Value} : Array.Empty<DiscordEmbed>(), default, default);
+            => this.ApiClient.EditMessageAsync(channel_id, message_id, default, embed.HasValue ? new[] {embed.Value} : Array.Empty<DiscordEmbed>(), default, default, default);
```
Comment (VelvetThePanda) [inline DSharpPlus.Rest/DiscordRestClient.cs:561]: `Array.Empty<T>`

File: DSharpPlus.Rest/DiscordRestClient.cs [hunk 8]
```diff
@@ -548,7 +548,7 @@ public Task<DiscordMessage> GetChannelMessageAsync(ulong channel_id, ulong messa
         /// <param name="content">New message content</param>
         /// <returns></returns>
         public Task<DiscordMessage> EditMessageAsync(ulong channel_id, ulong message_id, Optional<string> content)
-            => this.ApiClient.EditMessageAsync(channel_id, message_id, content, default, default, default);
+            => this.ApiClient.EditMessageAsync(channel_id, message_id, content, default, default, default, default);
```
Comment (VelvetThePanda) [inline DSharpPlus.Rest/DiscordRestClient.cs:551]: `Array.Empty<T>`

File: DSharpPlus/Net/Rest/DiscordApiClient.cs [hunk 9]
```diff
@@ -1063,14 +1063,24 @@ internal async Task<DiscordMessage> EditMessageAsync(ulong channel_id, ulong mes
...
             var url = Utilities.GetApiUriFor(path);
-            var res = await this.DoRequestAsync(this.Discord, bucket, url, RestRequestMethod.PATCH, route, payload: DiscordJson.SerializeObject(pld)).ConfigureAwait(false);
+            var res = await this.DoMultipartAsync(this.Discord, bucket, url, RestRequestMethod.PATCH, route, values: values, files: files ?? new List<DiscordMessageFile>()).ConfigureAwait(false);
 
             var ret = this.PrepareMessage(JObject.Parse(res.Response));
 
+            foreach (var file in files.Where(x => x.ResetPositionTo.HasValue))
+            {
+                file.Stream.Position = file.ResetPositionTo.Value;
+            }
```
Comment (VelvetThePanda) [inline DSharpPlus/Net/Rest/DiscordApiClient.cs:1082]: Braces can be dropped

**Review by VelvetThePanda — CHANGES_REQUESTED** [review]

**Review by VelvetThePanda — APPROVED** [review]
LGTM

File: DSharpPlus/Net/Rest/DiscordApiClient.cs (hunk 2 above)
Comment (IDoEverything) [reply DSharpPlus/Net/Rest/DiscordApiClient.cs:1047]: `DoMultipartAsync` takes `IReadOnlyCollection`

**Review by Neuheit — APPROVED** [review]

**Pull request merged**